plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeMetadataSchemas.java

```diff
@@ -0,0 +1,349 @@
+        Types.MessageTypeBuilder builder = Types.buildMessage();
+        ImmutableMap.Builder<List<String>, Type> primitiveTypesBuilder = ImmutableMap.builder();
+        try {
+            stream(OBJECT_MAPPER.readTree(jsonSchema).get("fields").elements())
+                    .filter(fieldNode -> !partitionColumnNames.contains(fieldNode.get("name").asText())) //TODO check with mixed case
+                    .map(fieldNode -> buildType(fieldNode, typeManager, OPTIONAL, columnMappingMode, ImmutableList.of(), primitiveTypesBuilder::put))
+                    .forEach(builder::addField);
+        }
+        catch (JsonProcessingException e) {
+            throw new TrinoException(DELTA_LAKE_INVALID_SCHEMA, getLocation(e), "Failed to parse serialized schema: " + jsonSchema, e);
+        }
+
+        return new DeltaLakeParquetSchemaMapping(builder.named("trino_schema"), primitiveTypesBuilder.buildOrThrow());
+    }
+
+    private static org.apache.parquet.schema.Type buildType(
+            JsonNode fieldNode,
+            TypeManager typeManager,
+            org.apache.parquet.schema.Type.Repetition repetition,
+            DeltaLakeSchemaSupport.ColumnMappingMode columnMappingMode,
+            List<String> parent,
+            BiConsumer<List<String>, Type> primitiveTypesConsumer)
+    {
+        JsonNode typeNode = fieldNode.get("type");
+        OptionalInt fieldId = OptionalInt.empty();
+        String physicalName;
+
+        switch (columnMappingMode) {
+            case ID -> {
+                String columnMappingId = fieldNode.get("metadata").get("delta.columnMapping.id").asText();
+                verify(!isNullOrEmpty(columnMappingId), "id is null or empty");
+                fieldId = OptionalInt.of(Integer.parseInt(columnMappingId));
+                // Databricks stores column statistics with physical name
+                physicalName = fieldNode.get("metadata").get("delta.columnMapping.physicalName").asText();
+                verify(!isNullOrEmpty(physicalName), "physicalName is null or empty");
+            }
+            case NAME -> {
+                physicalName = fieldNode.get("metadata").get("delta.columnMapping.physicalName").asText();
+                verify(!isNullOrEmpty(physicalName), "physicalName is null or empty");
+            }
+            default -> {
+                physicalName = fieldNode.get("name").asText();
+            }
+        }
+
+        return buildType(typeNode, typeManager, repetition, physicalName, fieldId, columnMappingMode, parent, primitiveTypesConsumer);
+    }
+
+    private static org.apache.parquet.schema.Type buildType(
+            JsonNode typeNode,
+            TypeManager typeManager,
+            org.apache.parquet.schema.Type.Repetition repetition,
+            String name,
+            OptionalInt id,
+            DeltaLakeSchemaSupport.ColumnMappingMode columnMappingMode,
+            List<String> parent,
+            BiConsumer<List<String>, Type> primitiveTypesConsumer)
```

Comment:
Why not pass in the builder instead of `BiConsumer` ? Not sure why we would need a more generic parameter here.